controllers/node_label_module_version_reconciler.go

```diff
@@ -0,0 +1,257 @@
+	}
+}
+
+func (nlmvr *NodeLabelModuleVersionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	node := v1.Node{}
+
+	if err := nlmvr.client.Get(ctx, types.NamespacedName{Name: req.Name}, &node); err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not get node: %v", err)
+	}
+
+	logger := log.FromContext(ctx).WithValues("node name", node.Name)
+
+	// get node's label data per module
+	modulesVersionLabels := nlmvr.helperAPI.getLabelsPerModules(node.Labels)
+
+	// get module loader pods on the node
+	moduleLoaderPods, err := nlmvr.helperAPI.getModuleLoaderPods(ctx, node.Name)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could get module loader pods for the node %s: %v", node.Name, err)
+	}
+
+	// get the labels to add and delete for this specific node
+	reconLabelsRes := nlmvr.helperAPI.reconcileLabels(modulesVersionLabels, moduleLoaderPods)
+
+	logLabelsUpdateData(logger, reconLabelsRes)
+
+	// update the node with the labels
+	err = nlmvr.helperAPI.updateNodeLabels(ctx, node.Name, reconLabelsRes)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not label node %s with added/deleted labels: %v", node.Name, err)
+	}
+
+	return ctrl.Result{Requeue: reconLabelsRes.requeue}, nil
+}
+
+//go:generate mockgen -source=node_label_module_version_reconciler.go -package=controllers -destination=mock_node_label_module_version_reconciler.go nodeLabelModuleVersionHelperAPI
+
+type nodeLabelModuleVersionHelperAPI interface {
+	getLabelsPerModules(nodeLabels map[string]string) map[string]*modulesVersionLabels
+	getModuleLoaderPods(ctx context.Context, nodeName string) ([]v1.Pod, error)
+	reconcileLabels(modulesLabels map[string]*modulesVersionLabels, moduleLoaderPods []v1.Pod) *reconcileLabelsResult
+	updateNodeLabels(ctx context.Context, nodeName string, reconLabelsRes *reconcileLabelsResult) error
+}
+
+type nodeLabelModuleVersionHelper struct {
+	client client.Client
+}
+
+func newNodeLabelModuleVersionHelper(client client.Client) nodeLabelModuleVersionHelperAPI {
+	return &nodeLabelModuleVersionHelper{client: client}
+}
+
+func (nlmvha *nodeLabelModuleVersionHelper) getLabelsPerModules(nodeLabels map[string]string) map[string]*modulesVersionLabels {
+	labelsPerModule := map[string]*modulesVersionLabels{}
+	for key, value := range nodeLabels {
+		if utils.IsVersionLabel(key) {
+			namespace, name, err := utils.GetNamespaceNameFromVersionLabel(key)
+			if err != nil {
+				continue
+			}
+			mapKey := namespace + "-" + name
+			if labelsPerModule[mapKey] == nil {
+				labelsPerModule[mapKey] = &modulesVersionLabels{name: name, namespace: namespace}
+			}
+
+			if utils.IsModuleVersionLabel(key) {
+				labelsPerModule[mapKey].moduleVersionLabel = value
+			}
+			if utils.IsModuleLoaderVersionLabel(key) {
+				labelsPerModule[mapKey].moduleLoaderVersionLabel = value
+			}
+			if utils.IsDevicePluginVersionLabel(key) {
+				labelsPerModule[mapKey].devicePluginVersionLabel = value
+			}
```

Comment:
```suggestion
			if utils.IsModuleVersionLabel(key) {
				labelsPerModule[mapKey].moduleVersionLabel = value
			} else if utils.IsModuleLoaderVersionLabel(key) {
				labelsPerModule[mapKey].moduleLoaderVersionLabel = value
			} else if utils.IsDevicePluginVersionLabel(key) {
				labelsPerModule[mapKey].devicePluginVersionLabel = value
			}
```

Just for clarity to show that a label cannot be true on more than 1 entry.

Reply:
IMHO "if" without "else" is a much cleaner and readable code. The names of the functions are a clear indication of which label we are checking.

Reply:
I may think that a label can be both a moduleVersionLabel and also a moduleLoaderVersionLabel. If I wish to understand if multiple cases are possible I will need to check the functions content while having an `else` make it clear that each of of them can only fit a single entry.

Reply:
changed to switch